File: cses.fi/PlanetsAndKingdoms.cpp
/*
 * https://cses.fi/problemset/task/1683/
 */



// Base Question for SCCs using Kosaraju's algo (refer to Striver's video for explanation)
#include <bits/stdc++.h>
using namespace std;

vector<vector<int> > graph, revGraph;
vector<bool> vis;
vector<int> res;
stack<int> st;

void dfs1(int node) {
	vis[node] = 1;
	for (int neighbor : graph[node]) {
		if (!vis[neighbor])
			dfs1(neighbor);
	}
	st.push(node);
}

void dfs2(int node, int root) {
	vis[node] = 1;
	res[node] = root;
	for (int neighbor : revGraph[node]) {
		if (!vis[neighbor])
			dfs2(neighbor, root);
	}
}

int main() {
	int n, m, u, v;
	cin >> n >> m;
	graph.resize(n + 1);
	revGraph.resize(n + 1);
	vis.resize(n + 1, 0);
	res.resize(n + 1);
	while(m--) {
		cin >> u >> v;
		graph[u].push_back(v);
		revGraph[v].push_back(u);
	}
	for (int i = 1; i <= n; i++) {
		if (vis[i])
			continue;
		dfs1(i);
	}
	for (int i = 1; i <= n; i++)
	    vis[i] = 0;
	m = 0;
	while (!st.empty()) {
		v = st.top();
		st.pop();
		if (vis[v])
			continue;
		m += 1;
		dfs2(v, m);
	}
	cout << m << endl;
	for (int i = 1; i <= n; i++)
		cout << res[i] << ' ';
}

File: cses.fi/PlanetsAndKingdoms.java
/*
 * https://cses.fi/problemset/task/1683/
 */
import java.io.*;
import java.util.*;


public class PlanetsAndKingdoms {
	
	private static Map<Integer, List<Integer> > graph = new HashMap<>(), revGraph = new HashMap<>();
	private static Set<Integer> vis = new HashSet<>();
	private static Stack<Integer> stack = new Stack<>();
	private static int[] res;
	
	private static void dfs1(int node) {
		vis.add(node);
		for (int neighbor : graph.getOrDefault(node, Collections.emptyList())) {
			if (!vis.contains(neighbor))
				dfs1(neighbor);
		}
		stack.push(node);
	}
	
	private static void dfs2(int node, int root) {
		vis.add(node);
		res[node] = root;
		for (int neighbor : revGraph.getOrDefault(node, Collections.emptyList())) {
			if (!vis.contains(neighbor))
				dfs2(neighbor, root);
		}
	}
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String[] temp = br.readLine().trim().split(" ");
		int n = Integer.parseInt(temp[0]), m = Integer.parseInt(temp[1]), k = 0;
		res = new int[n + 1];
		while (m-- > 0) {
			temp = br.readLine().trim().split(" ");
			int u = Integer.parseInt(temp[0]), v = Integer.parseInt(temp[1]);
			graph.computeIfAbsent(u, k1 -> new ArrayList<>()).add(v);
			revGraph.computeIfAbsent(v, k1 -> new ArrayList<>()).add(u);
		}
		for (int i = 1; i <= n; i++) {
		    if (!vis.contains(i))
		        dfs1(i);
		}
		vis.clear();
		while (!stack.isEmpty()) {
			int node = stack.pop();
			if (vis.contains(node))
				continue;
			k += 1;
			dfs2(node, k);
		}
		System.out.println(k);
		for (int i = 1; i <= n; i++)
		    System.out.print(res[i] + " ");
	}
}
